tsl/src/fdw/data_node_scan_plan.c

```diff
@@ -674,6 +688,689 @@ push_down_group_bys(PlannerInfo *root, RelOptInfo *hyper_rel, Hyperspace *hs,
+		return false;
+
+	return true;
+}
+
+/*
+ * Check if the query performs a join between a hypertble (outer) and a reference
+ * table (inner) and the join type is a LEFT JOIN, an INNER JOIN, or an implicit
+ * join.
+ */
+static bool
+is_safe_to_pushdown_reftable_join(PlannerInfo *root, List *join_reference_tables,
+								  RangeTblEntry *outertableref, RangeTblEntry *innertableref,
+								  JoinType jointype)
+{
+	Query *query = root->parse;
+	List *fromList;
+
+	Assert(root != NULL);
+	Assert(outertableref != NULL);
+	Assert(innertableref != NULL);
+
+	/*
+	 * We only reach this code path if the outer table is a
+	 * distributed hypertable.
+	 */
+	Assert(ts_is_hypertable(outertableref->relid));
+
+	/*
+	 * We support pushing down of INNER and LEFT joins only.
+	 *
+	 * Constructing queries representing partitioned OUTER, SEMI and ANTI
+	 * joins is hard, hence not considered right now.
+	 */
+	if (jointype != JOIN_INNER && jointype != JOIN_LEFT)
+		return false;
+
+	/* Check that at least one reference table is defined. */
+	if (join_reference_tables == NIL)
+		return false;
+
+	/* Only queries with two tables and one join are supported. */
+	fromList = query->jointree->fromlist;
+	if (!is_inner_or_left_join_two_tables(fromList) && !is_implicit_join_two_tables(fromList))
+		return false;
+
+	/* Right table has to be a distributed hypertable */
+	if (!list_member_oid(join_reference_tables, innertableref->relid))
+		return false;
+
+	/* Join can be pushed down */
+	return true;
+}
+
+/*
+ * Assess whether the join between inner and outer relations can be pushed down
+ * to the foreign server. As a side effect, save information we obtain in this
+ * function to TsFdwRelInfo passed in.
+ */
+static void
+fdw_pushdown_foreign_join(PlannerInfo *root, RelOptInfo *joinrel, JoinType jointype,
+						  RelOptInfo *outerrel, RelOptInfo *innerrel, JoinPathExtraData *extra)
+{
+	TsFdwRelInfo *fpinfo;
+	TsFdwRelInfo *fpinfo_o;
+	TsFdwRelInfo *fpinfo_i;
+	ListCell *lc;
+	List *joinclauses;
+
+	/*
+	 * If either of the joining relations is marked as unsafe to pushdown, the
+	 * join can not be pushed down.
+	 */
+	fpinfo = fdw_relinfo_get(joinrel);
+	fpinfo_o = fdw_relinfo_get(outerrel);
+	fpinfo_i = fdw_relinfo_get(innerrel);
+
+	/* shouldn't happen in the current implementation */
+	if (!fpinfo_o || !fpinfo_o->pushdown_safe || !fpinfo_i || !fpinfo_i->pushdown_safe)
+		ereport(ERROR,
+				(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+				 errmsg("join pushdown failed, a partition is not marked as pushdown safe")));
+
+	/*
+	 * If joining relations have local conditions, those conditions are
+	 * required to be applied before joining the relations. Hence the join can
+	 * not be pushed down (shouldn't happen in the current implementation).
+	 */
+	if (fpinfo_o->local_conds || fpinfo_i->local_conds)
+		ereport(ERROR,
+				(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+				 errmsg("join pushdown failed, local conditions are present")));
```

Comment:
```suggestion
				 errmsg("join pushdown failed because local conditions are present")));
```

Reply:
Again, should we really fail as opposed to just producing a non-pushdown plan?

Reply:
Yes, looks like this shouldn't be an error. We generally have to test four cases: ec/non-ec join, with shippable/non-shippable functions. I have some examples in dist_param test, scroll down from here: https://github.com/timescale/timescaledb/blob/main/tsl/test/sql/dist_param.sql#L64

Reply:
@erimatnor Changed, the join is not pushed down in that case.

@akuzm Thank you for pointing out that shippable/non-shippable / EC,non-EC cases are not tested. The test cases are now added.